Add test of kubectl support for multiple resources

janetkuo opened the pull request:
#14775

Test `kubectl create -f` with multiple resource (rc + svc) using YAML (separated by ---) and JSON (List typed). 

cc @bgrant0607

janetkuo committed:
Add test of kubectl support for multiple resources

bgrant0607 commented:
cc @liggitt @deads2k

bgrant0607 commented:
I assume this fails with --validate=true?

bgrant0607 commented on hack/testdata/multi-resource.json:
```diff
@@ -0,0 +1,60 @@
+         },
+         "spec":{
+             "ports": [{
+                 "name": "",
```
For one port, we shouldn't need a name field.

JanetKuo replied:
Fixed

bgrant0607 commented on hack/testdata/multi-resource.json:
```diff
@@ -0,0 +1,60 @@
+         "metadata":{
+            "name":"mock",
+            "labels":{
+               "name":"mock"
```
I know this is only a test, but please don't copy the practice of using "name" as a label key:
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/config-best-practices.md

I suggest "app". (Here, and in all other labels and selectors.)

JanetKuo replied:
Fixed

bgrant0607 commented on hack/testdata/multi-resource.json:
```diff
@@ -0,0 +1,60 @@
+         }
+      },
+      {
+         "kind":"Service",
```
Please list the service first.

JanetKuo replied:
Fixed

bgrant0607 commented on hack/test-cmd.sh:
```diff
@@ -722,6 +722,41 @@ __EOF__
+  kube::test::get_object_assert services "{{range.items}}{{$id_field}}:{{end}}" 'kubernetes:'
+  # Command
+  # TODO: remove --validate=false when PR "Add validate support for list kind #14726" is merged
+  kubectl create -f hack/testdata/multi-resource.json --validate=false "${kube_flags[@]}"
```
Will the test fail if this command exits with a non-zero status code?

JanetKuo replied:
Yes. The error message looks like this:
```
error validating "hack/testdata/multi-resource.json": error validating data: couldn't find type: v1.List; if you choose to ignore these errors, turn validation off with --validate=false
!!! Error in hack/test-cmd.sh:751
  'kubectl create -f hack/testdata/multi-resource.json "${kube_flags[@]}"' exited with status 1
Call stack:
  1: hack/test-cmd.sh:751 runTests(...)
  2: hack/test-cmd.sh:868 main(...)
Exiting with status 1
```

feihujiang commented:
I think We should add one test or some tests for ReplicationControllerList, ServiceList etc.

I tested the following example, and if --validate=false, they could be created.

The json example is as follows, which is fetched directly by api, not by kubectl get.
~~~
{
  "kind": "ReplicationControllerList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/namespaces/default/replicationcontrollers",
    "resourceVersion": "322971"
  },
  "items": [
    {
      "metadata": {
        "name": "cassandra",
        "namespace": "default",
        "selfLink": "/api/v1/namespaces/default/replicationcontrollers/cassandra",
        "uid": "4104888f-674b-11e5-92b4-286ed488f788",
        "resourceVersion": "322870",
        "generation": 1,
        "creationTimestamp": "2015-09-30T08:14:21Z",
        "labels": {
          "name": "cassandra"
        }
      },
      "spec": {
        "replicas": 1,
        "selector": {
          "name": "cassandra"
        },
        "template": {
          "metadata": {
            "creationTimestamp": null,
            "labels": {
              "name": "cassandra"
            }
          },
          "spec": {
            "volumes": [
              {
                "name": "data",
                "emptyDir": {}
              }
            ],
            "containers": [
              {
                "name": "cassandra",
                "image": "gcr.io/google_containers/cassandra:v6",
                "command": [
                  "/run.sh"
                ],
                "ports": [
                  {
                    "name": "cql",
                    "containerPort": 9042,
                    "protocol": "TCP"
                  },
                  {
                    "name": "thrift",
                    "containerPort": 9160,
                    "protocol": "TCP"
                  }
                ],
                "env": [
                  {
                    "name": "MAX_HEAP_SIZE",
                    "value": "512M"
                  },
                  {
                    "name": "HEAP_NEWSIZE",
                    "value": "100M"
                  },
                  {
                    "name": "POD_NAMESPACE",
                    "valueFrom": {
                      "fieldRef": {
                        "apiVersion": "v1",
                        "fieldPath": "metadata.namespace"
                      }
                    }
                  }
                ],
                "resources": {
                  "limits": {
                    "cpu": "100m"
                  },
                  "requests": {
                    "cpu": "100m"
                  }
                },
                "volumeMounts": [
                  {
                    "name": "data",
                    "mountPath": "/cassandra_data"
                  }
                ],
                "terminationMessagePath": "/dev/termination-log",
                "imagePullPolicy": "IfNotPresent"
              }
            ],
            "restartPolicy": "Always",
            "terminationGracePeriodSeconds": 30,
            "dnsPolicy": "ClusterFirst"
          }
        }
      },
      "status": {
        "replicas": 1,
        "observedGeneration": 1
      }
    },
    {
      "metadata": {
        "name": "frontend",
        "namespace": "default",
        "selfLink": "/api/v1/namespaces/default/replicationcontrollers/frontend",
        "uid": "8508283a-674b-11e5-92b4-286ed488f788",
        "resourceVersion": "322943",
        "generation": 1,
        "creationTimestamp": "2015-09-30T08:16:15Z",
        "labels": {
          "name": "frontend"
        }
      },
      "spec": {
        "replicas": 3,
        "selector": {
          "name": "frontend"
        },
        "template": {
          "metadata": {
            "creationTimestamp": null,
            "labels": {
              "name": "frontend"
            }
          },
          "spec": {
            "containers": [
              {
                "name": "php-redis",
                "image": "gcr.io/google_samples/gb-frontend:v3",
                "ports": [
                  {
                    "containerPort": 80,
                    "protocol": "TCP"
                  }
                ],
                "env": [
                  {
                    "name": "GET_HOSTS_FROM",
                    "value": "dns"
                  }
                ],
                "resources": {},
                "terminationMessagePath": "/dev/termination-log",
                "imagePullPolicy": "IfNotPresent"
              }
            ],
            "restartPolicy": "Always",
            "terminationGracePeriodSeconds": 30,
            "dnsPolicy": "ClusterFirst"
          }
        }
      },
      "status": {
        "replicas": 3,
        "observedGeneration": 1
      }
    }
  ]
}
~~~

jlowdermilk commented:
This lgtm. Up to you if you want to add tests for ReplicationControllerList, ServiceList in this pr.

JanetKuo commented:
Based on https://github.com/kubernetes/kubernetes/issues/14775#issuecomment-144327349, added tests for kubectl create, get, describe, replace, label, annotate, and delete with multiple resources specified. Also added tests with json object concatenation. cc @liggitt @smarterclayton

We don't plan to support user operating on ReplicationControllerList or ServiceList directly. See https://github.com/kubernetes/kubernetes/issues/14775#issuecomment-144496171.

JanetKuo commented on hack/test-cmd.sh:
```diff
@@ -722,6 +723,109 @@ __EOF__
+  # Command
+  # kubectl create -f hack/testdata/multi-resource.json "${kube_flags[@]}" # test fails here now
+  # TODO: test get when geting List is fixed
+  # kubectl get -f hack/testdata/multi-resource-list.json "${kube_flags[@]}"
```
When trying to get resources specified by JSON List, kubectl returns the following error. 
```
error: error: unknown type &api.List{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"", ResourceVersion:""}, Items:[]runtime.Object{(*runtime.Unknown)(0xc2080d2f40), (*runtime.Unknown)(0xc2080d3000)}}
```
It's because we don't do `Flatten()` in kubectl get before `Visit()`. Is it expected behavior or we should fix this? 
cc @smarterclayton @bgrant0607

jlowdermilk replied:
We should definitely fix that. I thought we already fixed get with List, but perhaps it regressed.

JanetKuo replied:
I'll fix it.

JanetKuo replied:
Fixing it in #14888

JanetKuo commented:
@k8s-bot test this please

bgrant0607 commented:
Ok, I'm surprised.
```
apiVersion: v1
kind: ReplicationControllerList
items:
 - apiVersion: v1
   kind: ReplicationController
   metadata:
     name: guestbook-redis-master
   spec:
     template:
       metadata:
         labels:
           app: guestbook
           tier: backend
           role: master
       spec:
         containers:
         - name: redis
           image: redis
           ports:
           - containerPort: 6379
 - apiVersion: v1
   kind: ReplicationController
   metadata:
     name: guestbook-redis-slave
   spec:
     replicas: 2
     template:
       metadata:
         labels:
           app: guestbook
           tier: backend
           role: slave
       spec:
         containers:
         - name: worker
           image: kubernetes/redis-slave:v2
           ports:
           - containerPort: 6379

$ cluster/kubectl.sh create -f redis-both.yaml
replicationcontroller "guestbook-redis-master" created
replicationcontroller "guestbook-redis-slave" created
$ cluster/kubectl.sh get -f !$
cluster/kubectl.sh get -f redis-both.yaml
CONTROLLER               CONTAINER(S)   IMAGE(S)                    SELECTOR                                 REPLICAS   AGE
guestbook-redis-master   redis          redis                       app=guestbook,role=master,tier=backend   1          <unknown>
guestbook-redis-slave    worker         kubernetes/redis-slave:v2   app=guestbook,role=slave,tier=backend    2          <unknown>
$ cluster/kubectl.sh get rc guestbook-redis-master
CONTROLLER               CONTAINER(S)   IMAGE(S)   SELECTOR                                 REPLICAS   AGE
guestbook-redis-master   redis          redis      app=guestbook,role=master,tier=backend   1          2m
$ cluster/kubectl.sh get rc guestbook-redis-slave
CONTROLLER              CONTAINER(S)   IMAGE(S)                    SELECTOR                                REPLICAS   AGE
guestbook-redis-slave   worker         kubernetes/redis-slave:v2   app=guestbook,role=slave,tier=backend   2          2m
```

Since it works now (except for get -f), we should test it.

bgrant0607 commented:
I'd rather have this in sooner rather than later. Additional tests can be added in subsequent PRs.

jlowdermilk commented:
@k8s-bot test this

bgrant0607 commented:
Please cherrypick this into 1.1 after it is merged.

JanetKuo commented:
Squashed commits.

JanetKuo commented:
@k8s-bot test this

JanetKuo commented:
@k8s-bot test this

JanetKuo commented:
@k8s-bot test this

smarterclayton commented on hack/test-cmd.sh:
```diff
@@ -727,6 +728,110 @@ __EOF__
+  kube::log::status "Testing kubectl(${version}:multiple resources)"
+  # TODO: add test for types like ReplicationControllerList, ServiceList
+
+  ### Create, get, describe, replace, label, annotate, and then delete service nginxsvc and replication controller my-nginx from YAML, separated by ---
```
As much as I dislike bash meta programming, why not make this a loop over three different file suffixes for each test and verify with the exact same steps?  for suffix in (list.json concat.json concat.yaml)

JanetKuo replied:
@smarterclayton thanks for the suggestion. Since it's merged I'll send another PR for this.
